README.md
# 🎨 Hue Map

[![npm version](https://img.shields.io/npm/v/hue-map)](https://www.npmjs.com/package/hue-map)
[![minzip size](https://img.shields.io/bundlephobia/minzip/hue-map)](https://bundlephobia.com/package/hue-map)

Interpolate gradients based on common palettes. **[See a demo](https://giraugh.github.io/hue-map/)**.

Inspired by D3 colour scales, with maps from [colormap](https://github.com/bpostlethwaite/colormap).

![Some example palettes](https://user-images.githubusercontent.com/8862273/185757207-643934bb-da49-42f6-9956-8185de6bcb19.png)

## Installation

```bash
yarn add hue-map
```

## Usage

```js
import { createPalette } from 'hue-map'

const myPalette = createPalette({
  map: 'viridis',
  steps: 3,
})

console.log(myPalette.format('cssHex'))
// --> ['#440154FF', '#21908DFF', '#FDE725FF']
```

## API

`hue-map` exports a `createPalette` function which takes an options object and returns a palette object.

### Palette Creation Options 

| Property | Type | Default | Description |
| -------- | ---- | ------- | ----------- |
| `map` | `string` | `'viridis'` | The colour map to use, see below for a list of built-in colour maps. You can also provide a custom colour map. |
| `steps` | `number` | `10` | The number of colour shades to return. |


### Palette Formatting

Call `.format()` on a palette object to get an array of colours. Pass a `format` argument to control the format that the colors are returned as.

| Format   | Returned Type | Example     | Description                                 |
| -------- | ------------- | ----------- | ------------------------------------------- |
| `cssHex` | `string`      | `'#F79E00FF'` | A valid css colour in the form of #RRGGBBAA |
| `cssRGBA` | `string`      | `'rgba(247, 158, 0, 1)'` | A valid css colour in the form of rgba(R, G, B, A) |
| `number` | `number`      | `0xF79E00FF` | A hex number representing the colour in the form of 0xRRGGBBAA |
| `float` | `[number, number, number, number]` | `[0.3, 0.2, 0.5, 1.0]`  | A four-tuple of numbers between 0 and 1 representing R, G, B, and A respectively |
| `rgba` | `[number, number, number, number]` | `[128, 40, 200, 255]` | A four-tuple of numbers between 0 and 255 representing R, G, B, and A respectively |


### Built-in maps

There are several built-in colour maps available to use:

`jet`, `hsv`, `hot`, `spring`, `summer`, `autumn`, `winter`, `bone`, `copper`, `greys`, `yignbu`, `greens`, `yiorrd`, `bluered`, `rdbu`, `picnic`, `rainbow`, `portland`, `blackbody`, `earth`, `electric`, `alpha`, `viridis`, `inferno`, `magma`, `plasma`, `warm`, `cool`, `rainbow-soft`, `bathymetry`, `cdom`, `chlorophyll`, `density`, `freesurface-blue`, `freesurface-red`, `oxygen`, `par`, `phase`, `salinity`, `temperature`, `turbidity`, `velocity-blue`, `velocity-green`, `cubehelix`

Visit the [demo page](https://giraugh.github.io/hue-map/) to see a list with examples.

### Custom colour maps

You can also provide a custom colour map to the `map` option, with a type of `[number, number | RGBA][]`. This is an array of tuples, where each tuple has an index of where that colour appears in the gradient (from 0 to 1), and the colour at that point, as a HEX number or an RGBA tuple. Note that all HEX numbers need to include alpha.

```js
import { createPalette } from 'hue-map'

// 3 colour points, using HEX values
const myHexPalette = createPalette({
  map: [
    [0, 0xFEAC5EFF],
    [0.5, 0xC779D0FF],
    [1, 0x4BC0C8FF],
  ]
})

// 5 colour points, using RGBA values
const myHexPalette = createPalette({
  map: [
    [0, [38, 83, 43, 255]],
    [0.25, [57, 158, 90, 255]],
    [0.5, [90, 188, 185, 255]],
    [0.75, [99, 226, 198, 255]],
    [1, [110, 249, 245, 255]],
  ]
})
```

## Development

1. Clone the repo onto your machine
2. Run `yarn` to install dependencies
3. Run `yarn dev` to test palettes with a demo page

Build the library to `dist` using `yarn build`.

## Contributing

Issue contributions and PRs are greatly welcomed and appreciated!

This library uses [changesets](https://github.com/changesets/changesets), if the changes you've made would constitute a version bump, run `yarn changeset` and follow the prompts to document the changes you've made. Changesets are consumed on releases, and used to generate a changelog and bump version number.

## License

`hue-map` is licensed under MIT


CHANGELOG.md
# hue-map

## 1.0.0

### Major Changes

- e400285: Refactor api to return a palette object which can be formatted later

## 0.1.3

### Patch Changes

- 5b129bd: Fix malformed module identifier in type declarations

## 0.1.2

### Patch Changes

- a891893: Output library types

## 0.1.1

### Patch Changes

- b9646fd: Fix interpolation of maps with more points than steps requested. Also ensures that the last colour is always the end-most colour from the map.

## 0.1.0

### Minor Changes

- e010332: Add support for providing custom color maps to the createPalette function


lib/maps.ts
import { RGBA } from './util'

const maps = {
  'jet':              [[0, 0x000083FF], [0.125, 0x003CAAFF], [0.375, 0x05FFFFFF], [0.625, 0xFFFF00FF], [0.875, 0xFA0000FF], [1, 0x800000FF]],
  'hsv':              [[0, 0xFF0000FF], [0.169, 0xFDFF02FF], [0.173, 0xF7FF02FF], [0.337, 0x00FC04FF], [0.341, 0x00FC0AFF], [0.506, 0x01F9FFFF], [0.671, 0x0200FDFF], [0.675, 0x0800FDFF], [0.839, 0xFF00FBFF], [0.843, 0xFF00F5FF], [1, 0xFF0006FF]],
  'hot':              [[0, 0x000000FF], [0.3, 0xE60000FF], [0.6, 0xFFD200FF], [1, 0xFFFFFFFF]],
  'spring':           [[0, 0xFF00FFFF], [1, 0xFFFF00FF]],
  'summer':           [[0, 0x008066FF], [1, 0xFFFF66FF]],
  'autumn':           [[0, 0xFF0000FF], [1, 0xFFFF00FF]],
  'winter':           [[0, 0x0000FFFF], [1, 0x00FF80FF]],
  'bone':             [[0, 0x000000FF], [0.376, 0x545474FF], [0.753, 0xA9C8C8FF], [1, 0xFFFFFFFF]],
  'copper':           [[0, 0x000000FF], [0.804, 0xFFA066FF], [1, 0xFFC77FFF]],
  'greys':            [[0, 0x000000FF], [1, 0xFFFFFFFF]],
  'yignbu':           [[0, 0x081D58FF], [0.125, 0x253494FF], [0.25, 0x225EA8FF], [0.375, 0x1D91C0FF], [0.5, 0x41B6C4FF], [0.625, 0x7FCDBBFF], [0.75, 0xC7E9B4FF], [0.875, 0xEDF8D9FF], [1, 0xFFFFD9FF]],
  'greens':           [[0, 0x00441BFF], [0.125, 0x006D2CFF], [0.25, 0x238B45FF], [0.375, 0x41AB5DFF], [0.5, 0x74C476FF], [0.625, 0xA1D99BFF], [0.75, 0xC7E9C0FF], [0.875, 0xE5F5E0FF], [1, 0xF7FCF5FF]],
  'yiorrd':           [[0, 0x800026FF], [0.125, 0xBD0026FF], [0.25, 0xE31A1CFF], [0.375, 0xFC4E2AFF], [0.5, 0xFD8D3CFF], [0.625, 0xFEB24CFF], [0.75, 0xFED976FF], [0.875, 0xFFEDA0FF], [1, 0xFFFFCCFF]],
  'bluered':          [[0, 0x0000FFFF], [1, 0xFF0000FF]],
  'rdbu':             [[0, 0x050AACFF], [0.35, 0x6A89F7FF], [0.5, 0xBEBEBEFF], [0.6, 0xDCAA84FF], [0.7, 0xE6915AFF], [1, 0xB20A1CFF]],
  'picnic':           [[0, 0x0000FFFF], [0.1, 0x3399FFFF], [0.2, 0x66CCFFFF], [0.3, 0x99CCFFFF], [0.4, 0xCCCCFFFF], [0.5, 0xFFFFFFFF], [0.6, 0xFFCCFFFF], [0.7, 0xFF99FFFF], [0.8, 0xFF66CCFF], [0.9, 0xFF6666FF], [1, 0xFF0000FF]],
  'rainbow':          [[0, 0x96005AFF], [0.125, 0x0000C8FF], [0.25, 0x0019FFFF], [0.375, 0x0098FFFF], [0.5, 0x2CFF96FF], [0.625, 0x97FF00FF], [0.75, 0xFFEA00FF], [0.875, 0xFF6F00FF], [1, 0xFF0000FF]],
  'portland':         [[0, 0x0C3383FF], [0.25, 0x0A88BAFF], [0.5, 0xF2D338FF], [0.75, 0xF28F38FF], [1, 0xD91E1EFF]],
  'blackbody':        [[0, 0x000000FF], [0.2, 0xE60000FF], [0.4, 0xE6D200FF], [0.7, 0xFFFFFFFF], [1, 0xA0C8FFFF]],
  'earth':            [[0, 0x000082FF], [0.1, 0x00B4B4FF], [0.2, 0x28D228FF], [0.4, 0xE6E632FF], [0.6, 0x784614FF], [1, 0xFFFFFFFF]],
  'electric':         [[0, 0x000000FF], [0.15, 0x1E0064FF], [0.4, 0x780064FF], [0.6, 0xA05A00FF], [0.8, 0xE6C800FF], [1, 0xFFFADCFF]],
  'alpha':            [[0, 0xFFFFFF00], [1, 0xFFFFFFFF]],
  'viridis':          [[0, 0x440154FF], [0.13, 0x472C7AFF], [0.25, 0x3B518BFF], [0.38, 0x2C718EFF], [0.5, 0x21908DFF], [0.63, 0x27AD81FF], [0.75, 0x5CC863FF], [0.88, 0xAADC32FF], [1, 0xFDE725FF]],
  'inferno':          [[0, 0x000004FF], [0.13, 0x1F0C48FF], [0.25, 0x550F6DFF], [0.38, 0x88226AFF], [0.5, 0xBA3655FF], [0.63, 0xE35933FF], [0.75, 0xF98C0AFF], [0.88, 0xF9C932FF], [1, 0xFCFFA4FF]],
  'magma':            [[0, 0x000004FF], [0.13, 0x1C1044FF], [0.25, 0x4F127BFF], [0.38, 0x812581FF], [0.5, 0xB5367AFF], [0.63, 0xE55064FF], [0.75, 0xFB8761FF], [0.88, 0xFEC287FF], [1, 0xFCFDBFFF]],
  'plasma':           [[0, 0x0D0887FF], [0.13, 0x4B03A1FF], [0.25, 0x7D03A8FF], [0.38, 0xA82296FF], [0.5, 0xCB4679FF], [0.63, 0xE56B5DFF], [0.75, 0xF89441FF], [0.88, 0xFDC328FF], [1, 0xF0F921FF]],
  'warm':             [[0, 0x7D00B3FF], [0.13, 0xAC00BBFF], [0.25, 0xDB00AAFF], [0.38, 0xFF0082FF], [0.5, 0xFF3F4AFF], [0.63, 0xFF7B00FF], [0.75, 0xEAB000FF], [0.88, 0xBEE400FF], [1, 0x93FF00FF]],
  'cool':             [[0, 0x7D00B3FF], [0.13, 0x7400DAFF], [0.25, 0x624AEDFF], [0.38, 0x4492E7FF], [0.5, 0x00CCC5FF], [0.63, 0x00F792FF], [0.75, 0x00FF58FF], [0.88, 0x28FF08FF], [1, 0x93FF00FF]],
  'rainbow-soft':     [[0, 0x7D00B3FF], [0.1, 0xC700B4FF], [0.2, 0xFF0079FF], [0.3, 0xFF6C00FF], [0.4, 0xDEC200FF], [0.5, 0x96FF00FF], [0.6, 0x00FF37FF], [0.7, 0x00F696FF], [0.8, 0x32A7DEFF], [0.9, 0x6733EBFF], [1, 0x7C00BAFF]],
  'bathymetry':       [[0, 0x281A2CFF], [0.13, 0x3B315AFF], [0.25, 0x404C8BFF], [0.38, 0x3F6E97FF], [0.5, 0x488E9EFF], [0.63, 0x55AEA3FF], [0.75, 0x78CEA3FF], [0.88, 0xBBE6ACFF], [1, 0xFDFECCFF]],
  'cdom':             [[0, 0x2F0F3EFF], [0.13, 0x571756FF], [0.25, 0x821C63FF], [0.38, 0xAB2960FF], [0.5, 0xCE4356FF], [0.63, 0xE66A54FF], [0.75, 0xF29567FF], [0.88, 0xF9C187FF], [1, 0xFEEDB0FF]],
  'chlorophyll':      [[0, 0x122414FF], [0.13, 0x193F29FF], [0.25, 0x185B3BFF], [0.38, 0x0D7748FF], [0.5, 0x129450FF], [0.63, 0x50AD59FF], [0.75, 0x84C47AFF], [0.88, 0xAFDDA2FF], [1, 0xD7F9D0FF]],
  'density':          [[0, 0x360E24FF], [0.13, 0x591750FF], [0.25, 0x6E2D84FF], [0.38, 0x784DB2FF], [0.5, 0x7871D5FF], [0.63, 0x7397E4FF], [0.75, 0x86B9E3FF], [0.88, 0xB1D6E3FF], [1, 0xE6F1F1FF]],
  'freesurface-blue': [[0, 0x1E046EFF], [0.13, 0x2F0EB0FF], [0.25, 0x292DECFF], [0.38, 0x1963D4FF], [0.5, 0x4483C8FF], [0.63, 0x729CC5FF], [0.75, 0x9DB5CBFF], [0.88, 0xC8D0D8FF], [1, 0xF1EDECFF]],
  'freesurface-red':  [[0, 0x3C0912FF], [0.13, 0x64111BFF], [0.25, 0x8E141DFF], [0.38, 0xB12B1BFF], [0.5, 0xC0573FFF], [0.63, 0xCD7D69FF], [0.75, 0xD8A294FF], [0.88, 0xE3C7C1FF], [1, 0xF1EDECFF]],
  'oxygen':           [[0, 0x400505FF], [0.13, 0x6A060FFF], [0.25, 0x901A07FF], [0.38, 0xA84003FF], [0.5, 0xBC6404FF], [0.63, 0xCE880BFF], [0.75, 0xDCAE19FF], [0.88, 0xE7D72CFF], [1, 0xF8FE69FF]],
  'par':              [[0, 0x331418FF], [0.13, 0x5A2023FF], [0.25, 0x812C22FF], [0.38, 0x9F4419FF], [0.5, 0xB66313FF], [0.63, 0xC78616FF], [0.75, 0xD4AB23FF], [0.88, 0xDDD236FF], [1, 0xE1FD4BFF]],
  'phase':            [[0, 0x916912FF], [0.13, 0xB84726FF], [0.25, 0xBA3A73FF], [0.38, 0xA047B9FF], [0.5, 0x6E61DAFF], [0.63, 0x327BA4FF], [0.75, 0x1F836EFF], [0.88, 0x4D8122FF], [1, 0x916912FF]],
  'salinity':         [[0, 0x2A186CFF], [0.13, 0x2132A2FF], [0.25, 0x0F5A91FF], [0.38, 0x287689FF], [0.5, 0x3B9287FF], [0.63, 0x4FAF7EFF], [0.75, 0x78CB68FF], [0.88, 0xC1DD64FF], [1, 0xFDEF9AFF]],
  'temperature':      [[0, 0x042333FF], [0.13, 0x17337AFF], [0.25, 0x553B9DFF], [0.38, 0x814F8FFF], [0.5, 0xAF5F82FF], [0.63, 0xDE7065FF], [0.75, 0xF99242FF], [0.88, 0xF9C441FF], [1, 0xE8FA5BFF]],
  'turbidity':        [[0, 0x221F1BFF], [0.13, 0x413229FF], [0.25, 0x624534FF], [0.38, 0x835939FF], [0.5, 0xA1703BFF], [0.63, 0xB98C42FF], [0.75, 0xCAAE58FF], [0.88, 0xD8D17EFF], [1, 0xE9F6ABFF]],
  'velocity-blue':    [[0, 0x112040FF], [0.13, 0x233474FF], [0.25, 0x1D519CFF], [0.38, 0x1F71A2FF], [0.5, 0x3290A9FF], [0.63, 0x57ADB0FF], [0.75, 0x95C4BDFF], [0.88, 0xCBDDD3FF], [1, 0xFEFBE6FF]],
  'velocity-green':   [[0, 0x172313FF], [0.13, 0x184026FF], [0.25, 0x0B5F2DFF], [0.38, 0x277B23FF], [0.5, 0x5F920CFF], [0.63, 0x98A512FF], [0.75, 0xC9BA45FF], [0.88, 0xE9D889FF], [1, 0xFFFDCDFF]],
  'cubehelix':        [[0, 0x000000FF], [0.07, 0x16053BFF], [0.13, 0x3C0469FF], [0.2, 0x6D0187FF], [0.27, 0xA10093FF], [0.33, 0xD2028EFF], [0.4, 0xFB0B7BFF], [0.47, 0xFF1D61FF], [0.53, 0xFF3645FF], [0.6, 0xFF552EFF], [0.67, 0xFF7822FF], [0.73, 0xFF9D25FF], [0.8, 0xF1BF39FF], [0.87, 0xE0DC5DFF], [0.93, 0xDAF18EFF], [1, 0xE3FDC6FF]],
}

export type ColorMap = [number, number | RGBA][]
export type MapKey = keyof typeof maps
export default maps as unknown as Record<string, ColorMap>
